cmd/promtool/analyze.go

```diff
@@ -0,0 +1,515 @@
+					}
+					return err
+				}
+				fmt.Fprintf(out, "%s=%v\n", seriesSel, *stats)
+				metastats.update(stats)
+			}
+		}
+	}
+	fmt.Println(metastats)
+	fmt.Println(metahistogram)
+	return nil
+}
+
+// Query the last_over_time(*_sum[duration]) series related to the matcher,
+// keeping the result small (avoids buckets).
+func queryLabelSets(ctx context.Context, api v1.API, metricName string, end time.Time, duration time.Duration) (model.Vector, error) {
+	query := fmt.Sprintf("last_over_time(%s[%s])", metricName, duration.String())
+
+	values, _, err := api.Query(ctx, query, end)
+	if err != nil {
+		return nil, err
+	}
+
+	vector, ok := values.(model.Vector)
+	if !ok {
+		return nil, fmt.Errorf("query for metrics resulted in non-Vector")
+	}
+	return vector, nil
+}
+
+func seriesSelector(bucketMetricName string, duration time.Duration) string {
+	builder := strings.Builder{}
+	builder.WriteString(bucketMetricName)
+	builder.WriteRune('[')
+	builder.WriteString(duration.String())
+	builder.WriteRune(']')
+
+	return builder.String()
+}
+
+func seriesSelectorWithLabels(bucketMetricName string, lbs model.LabelSet, duration time.Duration) string {
+	builder := strings.Builder{}
+	builder.WriteString(bucketMetricName)
+	builder.WriteRune('{')
+	first := true
+	for l, v := range lbs {
+		if first {
+			first = false
+		} else {
+			builder.WriteRune(',')
+		}
+		builder.WriteString(string(l))
+		builder.WriteString("=\"")
+		builder.WriteString(string(v))
+		builder.WriteRune('"')
+	}
+	builder.WriteRune('}')
+	builder.WriteRune('[')
+	builder.WriteString(duration.String())
+	builder.WriteRune(']')
+
+	return builder.String()
+}
+
+func querySamples(ctx context.Context, api v1.API, query string, end time.Time) (model.Matrix, error) {
+	values, _, err := api.Query(ctx, query, end)
+	if err != nil {
+		return nil, err
+	}
+
+	matrix, ok := values.(model.Matrix)
+	if !ok {
+		return nil, fmt.Errorf("query of buckets resulted in non-Matrix")
+	}
+
+	return matrix, nil
+}
+
+// minChanged/avgChanged/maxChanged is for the number of changed buckets.
+// minDelta/avgDelta/maxDelta is for the total absolute change
+// in the buckets across all timesteps.
+// minPop/avgPop/maxPop is for the number of populated (non-zero) buckets.
+// total is the total number of buckets across all samples in the series,
+// populated or not.
+type statistics struct {
+	minChanged, maxChanged, minDelta, maxDelta, minPop, maxPop, total int
+	avgChanged, avgDelta, avgPop                                      float64
+}
+
+func (s statistics) String() string {
+	return fmt.Sprintf("Bucket stats (min/avg/max) - number of buckets changed: %d/%.3f/%d total delta of changed buckets: %d/%.3f/%d number of populated buckets: %d/%.3f/%d. total number of buckets: %d", s.minChanged, s.avgChanged, s.maxChanged, s.minDelta, s.avgDelta, s.maxDelta, s.minPop, s.avgPop, s.maxPop, s.total)
+}
+
+func calcClassicBucketStatistics(matrix model.Matrix, histo *statshistogram) (*statistics, error) {
```

Comment:
So with the one-query approach I'm suggesting, this function has to look at the actual `Metric` in `matrix`. It has to group the samples by histograms, i.e. it has to ignore the `le` label and check for equality of everything else.